deepchecks/recommender/ranking.py

```diff
@@ -7,705 +7,423 @@
-            (2) or raise an exception.
-    Returns
-    ----------
-    prediction_coverage:
-        The prediction coverage of the recommendations as a percent
-        rounded to 2 decimal places
-    ----------
-    Metric Defintion:
-    Ge, M., Delgado-Battenfeld, C., & Jannach, D. (2010, September).
-    Beyond accuracy: evaluating recommender systems by coverage and serendipity.
-    In Proceedings of the fourth ACM conference on Recommender systems (pp. 257-260). ACM.
-    """
-    catalog = item_to_index.keys()
-    unique_items_catalog = set(catalog)
-    if len(catalog) != len(unique_items_catalog):
-        raise AssertionError("Duplicated items in catalog")
-
-    predicted_flattened = [p for sublist in recommendations for p in sublist]
-    unique_items_pred = set(predicted_flattened)
-
-    if not unique_items_pred.issubset(unique_items_catalog):
-        if unseen_warning:
-            warnings.warn("There are items in predictions but unseen in catalog. "
-                          "They are ignored from prediction_coverage calculation")
-            unique_items_pred = unique_items_pred.intersection(unique_items_catalog)
-        else:
-            raise AssertionError("There are items in predictions but unseen in catalog.")
-
-    num_unique_predictions = len(unique_items_pred)
-    prediction_coverage = round(num_unique_predictions / (len(catalog) * 1.0) * 100, 2)
-    return prediction_coverage
-
-# Cosine similarity code copied from https://github.com/talboger/fastdist, Licenced under MIT Licence,
-# Copyright (c) 2020 tal boger
-
-@jit(nopython=True, fastmath=True)
-def init_w(w, n):
-    """
-    :purpose:
-    Initialize a weight array consistent of 1s if none is given
-    This is called at the start of each function containing a w param
-    :params:
-    w      : a weight vector, if one was given to the initial function, else None
-             NOTE: w MUST be an array of np.float64. so, even if you want a boolean w,
-             convert it to np.float64 (using w.astype(np.float64)) before passing it to
-             any function
-    n      : the desired length of the vector of 1s (often set to len(u))
-    :returns:
-    w      : an array of 1s with shape (n,) if w is None, else return w un-changed
-    """
-    if w is None:
-        return np.ones(n)
-    else:
-        return w
 
 
-@jit(nopython=True, fastmath=True)
-def cosine(u, v, w=None):
-    """
-    :purpose:
-    Computes the cosine distance between two 1D arrays
-    NOTE: fastdist before v1.1.5 returned cosine similarity (which is 1 - distance)
-    :params:
-    u, v   : input arrays, both of shape (n,)
-    w      : weights at each index of u and v. array of shape (n,)
-             if no w is set, it is initialized as an array of ones
-             such that it will have no impact on the output
-    :returns:
-    cosine  : float, the cosine distance between u and v
+def recall_k(relevant_items: Union[List[int], np.array, None],
```

Comment:
regarding type - do we assume that items must be ints, or can be any hashable?

Reply:
You're right, the arguments should support a List of hashable, especially for INT / STRING.

If the items in the list are categorical in nature, such as movie titles or product names, it may be more natural and easier to use strings to represent them --> making the code more readable and intuitive, since it is easier to interpret the meaning of a string than an integer. 

On the other hand, strings involve more processing and memory allocation. For example, comparing strings requires comparing each character in the string, while comparing integers can be done more efficiently using machine-level operations.

In summary, we can either add more flexiblity by supporting Hashable type in general, or focus on pure performance optimization.

Reply:
Would prefer the flexible option